PR: https://github.com/odoo/odoo/pull/220383

From: 4bb4e08066449177f89382718ceadd840ce90d0e
From: JF Aubert
Date: 2025-08-27 02:38:32

Breaking data model changes scores: del:3 + add:4, change matches:
-    lot_producing_id = fields.Many2one(
+    lot_producing_ids = fields.Many2many(
+    lot_producing_ids = fields.Many2many('stock.lot', string='Lot/Serial Numbers', related='mo_id.lot_producing_ids')
-    finished_lot_id = fields.Many2one(
+    finished_lot_ids = fields.Many2many(
-    order_finished_lot_id = fields.Many2one('stock.lot', string="Finished Lot/Serial Number", related="raw_material_production_id.lot_producing_id")
+    order_finished_lot_ids = fields.Many2many('stock.lot', string="Finished Lot/Serial Number", related="raw_material_production_id.lot_producing_ids")

Total Changes: 248

[IMP] mrp[*]: Produce Multiple Serials in one MO

Allow Manufacturing Order to produce multiple serial numbers.

This change some fields to many2many in
mrp.production, mrp.workorder, stock.move, quality.check & quality.alert
and break the rule 1 mo = 1 serial.

Moreover, the MO's form view has been adapted:
* for 'lot' tracked products
  the lot can be selected or generated and is directly shown
  same for Shop Floor in 'Register Production / Serial'
* for 'serial' ones
  the serials can only be generated (by a wizard)
  and are visible using a smart button
  in Shop Floor, you are silently redirected to the wizard
There is also a Clear button to reset the producing lot/serials.

task: 4830458

See odoo/enterprise#90930
See odoo/upgrade#8117

closes odoo/odoo#220383

Signed-off-by: Steve Van Essche <svs@odoo.com>

================================= pseudo patch: =================================

--- a/addons/mrp/models/mrp_production.py
+++ b/addons/mrp/models/mrp_production.py
@@ -87,7 +87,7 @@ class MrpProduction(models.Model):
     product_uom_id = fields.Many2one(
         'uom.uom', 'Unit', domain="[('id', 'in', allowed_uom_ids)]",
         readonly=False, required=True, compute='_compute_uom_id', store=True, copy=True, precompute=True)
-    lot_producing_id = fields.Many2one(
+    lot_producing_ids = fields.Many2many(
         'stock.lot', string='Lot/Serial Number', copy=False,
         domain="[('product_id', '=', product_id)]", check_company=True)
     qty_producing = fields.Float(string="Quantity Producing", digits='Product Unit', copy=False)
@@ -271,6 +271,7 @@ class MrpProduction(models.Model):
         string='Date Category', store=False,
         search='_search_date_category', readonly=True
     )
+    serial_numbers_count = fields.Integer("Count of serial numbers", compute='_compute_serial_numbers_count')
 
     _name_uniq = models.Constraint(
         'unique(name, company_id)',
@@ -708,7 +709,7 @@ class MrpProduction(models.Model):
                     ('state', 'in', allowed_states),
                     ('product_qty', '>', 0),
                     ('location_id', 'in', wh_location_ids),
-                    ('raw_material_production_id', '!=', mo.id),
+                    ('raw_material_production_id', 'not in', mo.ids),
                     ('product_id', 'in', lines.product_id.ids),
                     '|', ('move_orig_ids', '=', False),
                         ('move_orig_ids', 'in', lines.ids)], limit=1):
@@ -847,27 +848,46 @@ class MrpProduction(models.Model):
             for date in dates
         )
 
-    @api.onchange('qty_producing', 'lot_producing_id')
-    def _onchange_producing(self):
+    @api.depends('lot_producing_ids')
+    def _compute_serial_numbers_count(self):
+        for production in self:
+            if production.product_tracking != 'serial':
+                production.serial_numbers_count = 0
+                continue
+            production.serial_numbers_count = len(production.lot_producing_ids)
+
+    def _change_producing(self):
         if self.state in ['draft', 'cancel'] or (self.state == 'done' and self.is_locked):
-            return
-        productions_bypass_qty_producting = self.filtered(lambda p: p.lot_producing_id and p.product_tracking == 'lot' and p._origin and p._origin.qty_producing == p.qty_producing)
+            return False
+        if self.product_tracking == 'serial' and self.lot_producing_ids:
+            self.qty_producing = len(self.lot_producing_ids)
+        productions_bypass_qty_producting = self.filtered(lambda p: p.lot_producing_ids and p.product_tracking == 'lot' and p._origin and p._origin.qty_producing == p.qty_producing)
         # sudo needed for portal users
         (self - productions_bypass_qty_producting).sudo()._set_qty_producing(False)
+        return True
+
+    @api.onchange('qty_producing')
+    def _onchange_qty_producing(self):
+        self._change_producing()
 
-    @api.onchange('lot_producing_id')
+    @api.onchange('lot_producing_ids')
     def _onchange_lot_producing(self):
-        res = self._can_produce_serial_number()
-        if res is not True:
-            return res
+        if self._change_producing():
+            res = self._can_produce_serial_numbers()
+            if res is not True:
+                return res
 
-    def _can_produce_serial_number(self, sn=None):
+    def _can_produce_serial_numbers(self, sns=None):
         self.ensure_one()
-        sn = sn or self.lot_producing_id
-        if self.product_id.tracking == 'serial' and sn and sn.id not in self.move_raw_ids.lot_ids.ids:
-            message, dummy = self.env['stock.quant'].sudo()._check_serial_number(self.product_id, sn, self.company_id)
-            if message:
-                return {'warning': {'title': _('Warning'), 'message': message}}
+        sns = sns or self.lot_producing_ids
+        if self.product_id.tracking == 'serial' and sns:
+            messages = []
+            for sn in sns:
+                message, _dummy = self.env['stock.quant'].sudo()._check_serial_number(self.product_id, sn, self.company_id)
+                if message:
+                    messages.append(message)
+            if messages:
+                return {'warning': {'title': _('Warning'), 'message': ','.join(messages)}}
         return True
 
     @api.constrains('move_finished_ids')
@@ -878,6 +898,12 @@ class MrpProduction(models.Model):
             if sum(order.move_byproduct_ids.filtered(lambda m: m.state != 'cancel').mapped('cost_share')) > 100:
                 raise ValidationError(_("The total cost share for a manufacturing order's by-products cannot exceed 100."))
 
+    @api.constrains('lot_producing_ids')
+    def _check_lot_producing_ids(self):
+        for record in self:
+            if record.product_tracking == 'lot' and len(record.lot_producing_ids) > 1:
+                raise UserError(_("You cannot set more than 1 lot"))
+
     def write(self, vals):
         if 'move_byproduct_ids' in vals and 'move_finished_ids' not in vals:
             vals['move_finished_ids'] = vals.get('move_finished_ids', []) + vals['move_byproduct_ids']
@@ -938,14 +964,10 @@ class MrpProduction(models.Model):
                 production.with_context(no_procurement=True)._autoconfirm_production()
                 if production in production_to_replan:
                     production._plan_workorders()
-            if production.state == 'done' and ('lot_producing_id' in vals or 'qty_producing' in vals):
+            if production.state == 'done' and 'qty_producing' in vals:
                 finished_move = production.move_finished_ids.filtered(
                     lambda move: move.product_id == production.product_id and move.state == 'done')
-                finished_move_lines = finished_move.move_line_ids
-                if 'lot_producing_id' in vals:
-                    finished_move_lines.write({'lot_id': vals.get('lot_producing_id')})
-                if 'qty_producing' in vals:
-                    finished_move.quantity = vals.get('qty_producing')
+                finished_move.quantity = vals.get('qty_producing')
             if self._has_workorders() and not production.workorder_ids.operation_id and vals.get('date_start') and not vals.get('date_finished'):
                 new_date_start = fields.Datetime.to_datetime(vals.get('date_start'))
                 if not production.date_finished or new_date_start >= production.date_finished:
@@ -1266,9 +1288,10 @@ class MrpProduction(models.Model):
     def _set_qty_producing(self, pick_manual_consumption_moves=True):
         if self.product_id.tracking == 'serial':
             qty_producing_uom = self.product_uom_id._compute_quantity(self.qty_producing, self.product_id.uom_id, rounding_method='HALF-UP')
+            qty_production_uom = self.product_uom_id._compute_quantity(self.product_qty, self.product_id.uom_id, rounding_method='HALF-UP')
             # allow changing a non-zero value to a 0 to not block mass produce feature
-            if qty_producing_uom != 1 and not (qty_producing_uom == 0 and self._origin.qty_producing != self.qty_producing):
-                self.qty_producing = self.product_id.uom_id._compute_quantity(1, self.product_uom_id, rounding_method='HALF-UP')
+            if qty_producing_uom != qty_production_uom and not (qty_producing_uom == 0 and self._origin.qty_producing != self.qty_producing):
+                self.qty_producing = self.product_id.uom_id._compute_quantity(len(self.lot_producing_ids), self.product_uom_id, rounding_method='HALF-UP')
 
         # waiting for a preproduction move before assignement
         is_waiting = self.warehouse_id.manufacture_steps != 'mrp_one_step' and self.picking_ids.filtered(lambda p: p.picking_type_id == self.warehouse_id.pbm_type_id and p.state not in ('done', 'cancel'))
@@ -1381,14 +1404,9 @@ class MrpProduction(models.Model):
         return (dest_moves | parent_moves).group_id.mrp_production_ids.filtered(lambda p: p.origin != self.origin) - self
 
     def set_qty_producing(self):
-        # This method is used to call `_set_lot_producing` when the onchange doesn't apply.
         self.ensure_one()
         self._set_qty_producing(False)
 
-    def _set_lot_producing(self):
-        self.ensure_one()
-        self.lot_producing_id = self.env['stock.lot'].create(self._prepare_stock_lot_values())
-
     def action_view_mrp_production_childs(self):
         self.ensure_one()
         mrp_production_ids = self._get_children().ids
@@ -1459,13 +1477,29 @@ class MrpProduction(models.Model):
             'name': name,
         }
 
-    def action_generate_serial(self):
+    def action_generate_serial(self, workorder=False):
         self.ensure_one()
-        self._set_lot_producing()
-        if self.product_id.tracking == 'serial':
-            self._set_qty_producing(False)
-        if self.picking_type_id.auto_print_generated_mrp_lot:
-            return self._autoprint_generated_lot(self.lot_producing_id)
+        if self.product_tracking == 'lot':
+            if self.lot_producing_ids:
+                raise UserError(_("You cannot set more than 1 lot per product"))
+            self.lot_producing_ids = [Command.create(self._prepare_stock_lot_values())]
+            if self.picking_type_id.auto_print_generated_mrp_lot:
+                return self._autoprint_generated_lot(self.lot_producing_ids[-1])
+        elif self.product_tracking == 'serial':
+            if self.product_qty == 1 and not self.lot_producing_ids:
+                self.lot_producing_ids = [Command.create(self._prepare_stock_lot_values())]
+                self.qty_producing = 1
+                (workorder or self).set_qty_producing()
+                if self.picking_type_id.auto_print_generated_mrp_lot:
+                    return self._autoprint_generated_lot(self.lot_producing_ids[-1])
+                return
+            action = self.env["ir.actions.actions"]._for_xml_id("mrp.action_assign_serial_numbers")
+            action['context'] = {
+                'default_production_id': self.id,
+            }
+            if workorder:
+                action['context']['default_workorder_id'] = workorder.id
+            return action
 
     def action_confirm(self):
         self._check_company()
@@ -1766,6 +1800,8 @@ class MrpProduction(models.Model):
             # the finish move can already be completed by the workorder.
             for move in finish_moves:
                 move.quantity = order.product_uom_id.round(order.qty_producing - order.qty_produced, rounding_method='HALF-UP')
+                if move.has_tracking != 'none' and not move.lot_ids:
+                    move.lot_ids = order.lot_producing_ids.ids
                 extra_vals = order._prepare_finished_extra_vals()
                 if extra_vals:
                     move.move_line_ids.write(extra_vals)
@@ -1806,7 +1842,7 @@ class MrpProduction(models.Model):
             'procurement_group_id': self.procurement_group_id.id,
             'move_raw_ids': None,
             'move_finished_ids': None,
-            'lot_producing_id': False,
+            'lot_producing_ids': False,
             'origin': self.origin,
             'state': 'draft' if self.state == 'draft' else 'confirmed',
             'date_deadline': self.date_deadline,
@@ -2176,6 +2212,7 @@ class MrpProduction(models.Model):
     def pre_button_mark_done(self):
         self._button_mark_done_sanity_checks()
         production_auto_ids = set()
+        production_missing_lot_ids = set()
         for production in self:
             if not production.product_uom_id.is_zero(production.qty_producing):
                 production.move_raw_ids.filtered(
@@ -2184,8 +2221,13 @@ class MrpProduction(models.Model):
                 continue
             if production._auto_production_checks():
                 production_auto_ids.add(production.id)
-            else:
-                return production.action_mass_produce()
+            elif not production.lot_producing_ids:
+                production_missing_lot_ids.add(production.id)
+
+        if production_missing_lot_ids:
+            if len(production_missing_lot_ids) > 1:
+                raise UserError(_("You need to generate Lot/Serial Number(s) to mark as done some productions"))
+            return self.env['mrp.production'].browse(production_missing_lot_ids).action_generate_serial()
 
         productions_auto = self.env['mrp.production'].browse(production_auto_ids)
         for production in productions_auto:
@@ -2334,7 +2376,7 @@ class MrpProduction(models.Model):
             'view_id': self.env.ref('mrp.mrp_unbuild_form_view_simplified').id,
             'type': 'ir.actions.act_window',
             'context': {'default_product_id': self.product_id.id,
-                        'default_lot_id': self.lot_producing_id.id,
+                        'default_lot_id': self.lot_producing_ids[:1].id,
                         'default_mo_id': self.id,
                         'default_company_id': self.company_id.id,
                         'default_location_id': self.location_dest_id.id,
@@ -2343,19 +2385,6 @@ class MrpProduction(models.Model):
             'target': 'new',
         }
 
-    def action_mass_produce(self):
-        self.ensure_one()
-        self._check_company()
-        if self.state not in ['draft', 'confirmed', 'progress', 'to_close'] or\
-                self._auto_production_checks():
-            return
-
-        action = self.env["ir.actions.actions"]._for_xml_id("mrp.action_mrp_batch_produce")
-        action['context'] = {
-            'default_production_id': self.id,
-        }
-        return action
-
     def action_split(self):
         self._pre_action_split_merge_hook(split=True)
         if len(self) > 1:
@@ -2618,9 +2647,10 @@ class MrpProduction(models.Model):
     def _check_sn_uniqueness(self):
         """ Alert the user if the serial number as already been consumed/produced """
         self.ensure_one()
-        if self.product_tracking == 'serial' and self.lot_producing_id and self.lot_producing_id.id not in self.move_raw_ids.lot_ids.ids:
-            if self._is_finished_sn_already_produced(self.lot_producing_id):
-                raise UserError(_('This serial number for product %s has already been produced', self.product_id.name))
+        if self.product_tracking == 'serial' and self.lot_producing_ids:
+            lots_to_check = self.lot_producing_ids.filtered(lambda l: l.id not in self.move_raw_ids.lot_ids.ids)
+            if lots_to_check and self._are_finished_serials_already_produced(lots_to_check):
+                raise UserError(_('Serial number(s) for product %(product_name)s already produced', product_name=self.product_id.name))
 
         for move in self.move_finished_ids:
             if move.has_tracking != 'serial' or move.product_id == self.product_id:
@@ -2628,7 +2658,7 @@ class MrpProduction(models.Model):
             for move_line in move.move_line_ids:
                 if move_line.product_uom_id.is_zero(move_line.quantity):
                     continue
-                if self._is_finished_sn_already_produced(move_line.lot_id, excluded_sml=move_line):
+                if self._are_finished_serials_already_produced(move_line.lot_id, excluded_sml=move_line):
                     raise UserError(_('The serial number %(number)s used for byproduct %(product_name)s has already been produced',
                                       number=move_line.lot_id.name, product_name=move_line.product_id.name))
 
@@ -2685,12 +2715,12 @@ class MrpProduction(models.Model):
             if consumed_qty - cancelled_qty > 0:
                 raise UserError(sn_error_msg[sn_id])
 
-    def _is_finished_sn_already_produced(self, lot, excluded_sml=None):
-        if not lot:
+    def _are_finished_serials_already_produced(self, lots, excluded_sml=None):
+        if not lots:
             return False
         excluded_sml = excluded_sml or self.env['stock.move.line']
         domain = [
-            ('lot_id', '=', lot.id),
+            ('lot_id', 'in', lots.ids),
             ('quantity', '=', 1),
             ('state', '=', 'done')
         ]
@@ -2710,13 +2740,13 @@ class MrpProduction(models.Model):
                 ('move_id.unbuild_id', '!=', False)
             ])
             removed = self.env['stock.move.line'].search_count([
-                ('lot_id', '=', lot.id),
+                ('lot_id', 'in', lots.ids),
                 ('state', '=', 'done'),
                 ('location_id.scrap_location', '=', False),
                 ('location_dest_id.scrap_location', '=', True),
             ])
             unremoved = self.env['stock.move.line'].search_count([
-                ('lot_id', '=', lot.id),
+                ('lot_id', 'in', lots.ids),
                 ('state', '=', 'done'),
                 ('location_id.scrap_location', '=', True),
                 ('location_dest_id.scrap_location', '=', False),
@@ -2725,7 +2755,7 @@ class MrpProduction(models.Model):
             if not ((duplicates_unbuild or removed) and duplicates - duplicates_unbuild - removed + unremoved == 0):
                 return True
         # Check presence of same sn in current production
-        duplicates = co_prod_move_lines.filtered(lambda ml: ml.quantity and ml.lot_id == lot)
+        duplicates = co_prod_move_lines.filtered(lambda ml: ml.quantity and ml.lot_id.id in lots.ids)
         return bool(duplicates)
 
     def _pre_action_split_merge_hook(self, merge=False, split=False):
@@ -2770,12 +2800,9 @@ class MrpProduction(models.Model):
     def _set_quantities(self):
         self.ensure_one()
         missing_lot_id_products = ""
-        if self.product_tracking in ('lot', 'serial') and not self.lot_producing_id:
+        if self.product_tracking in ('lot', 'serial') and not self.lot_producing_ids:
             self.action_generate_serial()
-        if self.product_tracking == 'serial' and self.product_uom_id.compare(self.qty_producing, 1) == 1:
-            self.qty_producing = 1
-        else:
-            self.qty_producing = self.product_qty - self.qty_produced
+        self.qty_producing = self.product_qty - self.qty_produced
         self._set_qty_producing()
         self._mark_byproducts_as_produced()
 
@@ -2866,8 +2893,6 @@ class MrpProduction(models.Model):
 
     def _prepare_finished_extra_vals(self):
         self.ensure_one()
-        if self.lot_producing_id:
-            return {'lot_id' : self.lot_producing_id.id}
         return {}
 
     def action_open_label_layout(self):
@@ -2903,6 +2928,21 @@ class MrpProduction(models.Model):
         if self.state == "confirmed":
             self.state = "progress"
 
+    def action_view_serial_numbers(self):
+        action = self.env["ir.actions.actions"]._for_xml_id("stock.action_production_lot_form")
+        action['domain'] = [('id', 'in', self.lot_producing_ids.ids)]
+        action['name'] = _("Serial Numbers")
+        action['context'] = {
+            'create': False,
+            'delete': False,
+        }
+        return action
+
+    def action_clear_lot_producing_ids(self):
+        self.lot_producing_ids = [Command.clear()]
+        self.qty_producing = 0
+        self._set_qty_producing(False)
+
     def _track_subtype(self, init_values):
         self.ensure_one()
         if 'state' in init_values and self.state == 'confirmed':

--- a/addons/mrp/models/mrp_unbuild.py
+++ b/addons/mrp/models/mrp_unbuild.py
@@ -53,10 +53,10 @@ class MrpUnbuild(models.Model):
         domain="[('state', '=', 'done'), ('product_id', '=?', product_id), ('bom_id', '=?', bom_id)]",
         check_company=True, index='btree_not_null')
     mo_bom_id = fields.Many2one('mrp.bom', 'Bill of Material used on the Production Order', related='mo_id.bom_id')
+    lot_producing_ids = fields.Many2many('stock.lot', string='Lot/Serial Numbers', related='mo_id.lot_producing_ids')
     lot_id = fields.Many2one(
         'stock.lot', 'Lot/Serial Number',
-        compute='_compute_lot_id', store=True,
-        domain="[('product_id', '=', product_id)]", check_company=True)
+        domain="[('product_id', '=', product_id),('id', 'in', lot_producing_ids)]", check_company=True)
     has_tracking = fields.Selection(related='product_id.tracking', readonly=True)
     location_id = fields.Many2one(
         'stock.location', 'Source Location',
@@ -113,12 +113,6 @@ class MrpUnbuild(models.Model):
                     order.product_id, company_id=order.company_id.id
                 )[order.product_id]
 
-    @api.depends('mo_id')
-    def _compute_lot_id(self):
-        for order in self:
-            if order.mo_id:
-                order.lot_id = order.mo_id.lot_producing_id
-
     @api.depends('mo_id')
     def _compute_product_id(self):
         for order in self:

--- a/addons/mrp/models/mrp_workorder.py
+++ b/addons/mrp/models/mrp_workorder.py
@@ -111,8 +111,8 @@ class MrpWorkorder(models.Model):
     move_line_ids = fields.One2many(
         'stock.move.line', 'workorder_id', 'Moves to Track',
         help="Inventory moves for which you must scan a lot number at this work order")
-    finished_lot_id = fields.Many2one(
-        'stock.lot', string='Lot/Serial Number', related='production_id.lot_producing_id',
+    finished_lot_ids = fields.Many2many(
+        'stock.lot', string='Lot/Serial Numbers', related='production_id.lot_producing_ids',
         domain="[('product_id', '=', product_id), ('company_id', '=', company_id)]",
         readonly=False, check_company=True)
     time_ids = fields.One2many(
@@ -470,10 +470,10 @@ class MrpWorkorder(models.Model):
         )
         return interval['hours'] * 60
 
-    @api.onchange('finished_lot_id')
-    def _onchange_finished_lot_id(self):
+    @api.onchange('finished_lot_ids')
+    def _onchange_finished_lot_ids(self):
         if self.production_id:
-            res = self.production_id._can_produce_serial_number(sn=self.finished_lot_id)
+            res = self.production_id._can_produce_serial_numbers(sns=self.finished_lot_ids)
             if res is not True:
                 return res
 
@@ -485,8 +485,6 @@ class MrpWorkorder(models.Model):
                 raise UserError(_('You cannot change the quantity produced of a work order that is in done or cancel state.'))
             elif self.product_uom_id.compare(values['qty_produced'], 0) < 0:
                 raise UserError(_('The quantity produced must be positive.'))
-            elif values['qty_produced'] not in (0, 1) and any(wo.product_tracking == 'serial' for wo in self):
-                raise UserError(_('You cannot produce more than 1 unit of a serial product at a time.'))
 
         if 'production_id' in values and any(values['production_id'] != w.production_id.id for w in self):
             raise UserError(_('You cannot link this work order to another manufacturing order.'))
@@ -649,9 +647,7 @@ class MrpWorkorder(models.Model):
                     continue
                 raise UserError(_('You cannot start a work order that is already done or cancelled'))
 
-            if wo.product_tracking == 'serial' and wo.qty_producing == 0:
-                wo.qty_producing = 1.0
-            elif wo.qty_producing == 0:
+            if wo.qty_producing == 0:
                 wo.qty_producing = wo.qty_remaining
 
             if wo._should_start_timer():
@@ -878,42 +874,6 @@ class MrpWorkorder(models.Model):
             'company_id': self.company_id.id,
         }
 
-    def _update_finished_move(self):
-        """ Update the finished move & move lines in order to set the finished
-        product lot on it as well as the produced quantity. This method get the
-        information either from the last workorder or from the Produce wizard."""
-        production_move = self.production_id.move_finished_ids.filtered(
-            lambda move: move.product_id == self.product_id and
-            move.state not in ('done', 'cancel')
-        )
-        if not production_move:
-            return
-        if production_move.product_id.tracking != 'none':
-            if not self.finished_lot_id:
-                raise UserError(_('You need to provide a lot for the finished product.'))
-            move_line = production_move.move_line_ids.filtered(
-                lambda line: line.lot_id.id == self.finished_lot_id.id
-            )
-            if move_line:
-                if self.product_id.tracking == 'serial':
-                    raise UserError(_('You cannot produce the same serial number twice.'))
-                move_line.picked = True
-                move_line.quantity += self.qty_producing
-            else:
-                quantity = self.product_uom_id._compute_quantity(self.qty_producing, self.product_id.uom_id, rounding_method='HALF-UP')
-                putaway_location = production_move.location_dest_id._get_putaway_strategy(self.product_id, quantity)
-                move_line.create({
-                    'move_id': production_move.id,
-                    'product_id': production_move.product_id.id,
-                    'lot_id': self.finished_lot_id.id,
-                    'product_uom_id': self.product_uom_id.id,
-                    'quantity': self.qty_producing,
-                    'location_id': production_move.location_id.id,
-                    'location_dest_id': putaway_location.id,
-                })
-        else:
-            production_move.quantity = production_move.product_uom.round(self.qty_producing)
-
     def _should_start_timer(self):
         return True
 

--- a/addons/mrp/models/stock_move.py
+++ b/addons/mrp/models/stock_move.py
@@ -31,7 +31,7 @@ class StockMove(models.Model):
         'mrp.bom.byproduct', 'By-products', check_company=True,
         help="By-product line that generated the move in a manufacturing order")
     unit_factor = fields.Float('Unit Factor', compute='_compute_unit_factor', store=True)
-    order_finished_lot_id = fields.Many2one('stock.lot', string="Finished Lot/Serial Number", related="raw_material_production_id.lot_producing_id")
+    order_finished_lot_ids = fields.Many2many('stock.lot', string="Finished Lot/Serial Number", related="raw_material_production_id.lot_producing_ids")
     should_consume_qty = fields.Float('Quantity To Consume', compute='_compute_should_consume_qty', digits='Product Unit')
     cost_share = fields.Float(
         "Cost Share (%)", digits=(5, 2),  # decimal = 2 is important for rounding calculations!!
@@ -491,8 +491,8 @@ class StockMove(models.Model):
         vals = super()._prepare_move_line_vals(quantity, reserved_quant)
         if self.raw_material_production_id:
             vals['production_id'] = self.raw_material_production_id.id
-        if self.production_id.product_tracking == 'lot' and self.product_id == self.production_id.product_id:
-            vals['lot_id'] = self.production_id.lot_producing_id.id
+        if self.production_id.product_tracking == 'lot' and self.product_id == self.production_id.product_id and self.production_id.lot_producing_ids:
+            vals['lot_id'] = self.production_id.lot_producing_ids.ids[0]
         return vals
 
     def _key_assign_picking(self):

--- a/addons/mrp/models/stock_move_line.py
+++ b/addons/mrp/models/stock_move_line.py
@@ -34,7 +34,7 @@ class StockMoveLine(models.Model):
             # traceability report
             if line.move_id.raw_material_production_id and line.state == 'done':
                 mo = line.move_id.raw_material_production_id
-                finished_lots = mo.lot_producing_id
+                finished_lots = mo.lot_producing_ids
                 finished_lots |= mo.move_finished_ids.filtered(lambda m: m.product_id != mo.product_id).move_line_ids.lot_id
                 if finished_lots:
                     produced_move_lines = mo.move_finished_ids.move_line_ids.filtered(lambda sml: sml.lot_id in finished_lots)
